### DISCUSSION TITLE
Feature request: allow `visual-line-mode` to behave correctly in vertico minibuffers

### DISCUSSION THREAD

**Discussion by hab25** [minad/vertico#476] [Ideas]
Why do I want this: I use `global-visual-line-mode` and am not accostumed to, nor do I know or wish to learn the keyboard keybinding for, horizontal scrolling. However, I also use `marginalia` and want to see entire anotations without having to horizontally scroll, e.g. the entire documentation of some function in the minibuffer of `describe-functions`.

Currently, it is not possible (without advice, which is much less future-resistant) to use `visual-line-mode` normally in the vertico minibuffer, because it sets `truncate-lines`:

https://github.com/minad/vertico/blob/1def56a1e678ff9a8eeafafda9c474a5fbb097c4/vertico.el#L613

N.b., the above gets indirectly called by `post-command-hook` so one cannot use e.g. `minibuffer-setup-hook` to e.g. turn-off-then-on `visual-line-mode` such that it overrides with its own value of `truncate-lines`.

N.b., as can be inferred with the above linked code line, if one moves point far-enough beyond the window width e.g. by giving a long-enough prompt, then `(setq-local truncate-lines nil)` happens. This is what `visual-line-mode` expects and so it seems works correctly in that niche case.

Therefore, a simple code change like wrapping the linked line with `(unless visual-line-mode ...)` might be enough.

---

Bonus changes:

- Note that, when `(setq-local truncate-lines nil)` happens, the height of the vertico window, which is what I expect is what most users are really trying to control when they set `vertico-count`, is no longer strictly correlated with `vertico-count`, because each candidate could span accross a number of lines that is individual to each candidate. So, perhaps renaming it to `vertico-height` and having it alternatively accept a float indicating the proportion of the frame's height that the vertico window should occupy would be a useful change to go along with this. Such a sibling change would likely complicate current code that needs to calculate with `vertico-count`, but that would be worthwhile in my leyman's opinion; I expect the additional complication would small enough.

- Finally, specific to marginalia, another change could be done there or in some new `vertico-marginalia.el` to support this: have subsequent lines of marginalia anotations that wrap be indented to the starting column of the marginalia anotation, for ease of reading.

**Comment by minad** [comment]
Thanks for the request. A few comments:

> Therefore, a simple code change like wrapping the linked line with (unless visual-line-mode ...) might be enough.

Yes, this could work. Alternatively, it would be cleaner to introduce a separate customization variable controlling truncation/wrapping in Vertico. However we should not add such a check until the height computation gets improved. For now you can use an advice.

> Such a sibling change would likely complicate current code that needs to calculate with vertico-count, but that would be worthwhile in my leyman's opinion; I expect the additional complication would small enough.

Unfortunately it is not a small or simple change - it is the main complication. However with such a computation we could also support multi-line candidates better (candidates with \\n).

> Finally, specific to marginalia, another change could be done there or in some new vertico-marginalia.el to support this: have subsequent lines of marginalia annotations that wrap be indented to the starting column of the marginalia anotation, for ease of reading.

This is out of scope since I don't want to couple the packages tightly.

**Reply by hab25** [reply]
> Alternatively, it would be cleaner to introduce a separate customization variable controlling truncation/wrapping in Vertico.

Agreed, and that would also probably be a better interface for users because AFAICT there is currently nothing like a `vertico-setup-hook` that would allow a user to enable or disable `visual-line-mode` `vertico`-specifically. (Though, as a workaround, one can simulate that by combining `minibuffer-setup-hook` with `(when (eq (current-local-map) vertico-map) ...)`)

> This is out of scope since I don't want to couple the packages tightly.

Fair.

Thank you!

**Comment by hab25** [comment]
Another good reason to want this: avoiding truncation in `consult-ripgrep` and `consult-line`'s candidates.

**Reply by minad** [reply]
The main reason to truncate in `consult-ripgrep` and `consult-grep` is performance (`consult-grep-max-columns`). Otherwise Emacs would get overloaded quickly. For `consult-line` there was the idea to instead shift the line to the match position, while keeping single lines, see [vertico-truncate](https://github.com/jdtsmith/vertico-truncate).
